modules/distributor/usage/tracker.go

```diff
@@ -0,0 +1,423 @@
+package usage
+
+import (
+	"maps"
+	"math"
+	"math/bits"
+	"net/http"
+	"slices"
+	"sync"
+	"time"
+
+	"github.com/cespare/xxhash/v2"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/prometheus/prometheus/util/strutil"
+
+	v1 "github.com/grafana/tempo/pkg/tempopb/trace/v1"
+)
+
+const (
+	tenantLabel   = "tenant"
+	trackerLabel  = "tracker"
+	missingLabel  = "__missing__"
+	overflowLabel = "__overflow__"
+)
+
+type (
+	tenantLabelsFunc func(string) map[string]string
+	tenantMaxFunc    func(string) uint64
+)
+
+type bucket struct {
+	// Configuration
+	descr  *prometheus.Desc // Configuration can change over time so it is captured with the bucket.
+	labels []string
+
+	// Runtime data
+	bytes       uint64
+	lastUpdated int64
+}
+
+func (b *bucket) Inc(bytes uint64, unix int64) {
+	b.bytes += bytes
+	b.lastUpdated = unix
```

Comment:
Maybe we could skip passing the Unix timestamp from the outside

b.lastUpdated = time.Now().Unix()

Reply:
This is effectively called for every span.  `time.Now()` is fairly fast but it's not free, and this is more efficient.